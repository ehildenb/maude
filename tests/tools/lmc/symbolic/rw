#!/bin/sh

MAUDE_LIB=$srcdir/../../../src/Main
export MAUDE_LIB

../../../src/Main/maude \
  < $srcdir/rw.maude -no-banner -no-advise \
  > rw.out 2>&1

diff $srcdir/rw.expected rw.out > /dev/null 2>&1
