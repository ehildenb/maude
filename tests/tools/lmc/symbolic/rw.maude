
load symbolic-checker

--- The readers-writers problem in Maude
--- From: All About Maude - A High-Performance Logical Framework.
(mod READERS-WRITERS is 
  sort Nat . 
  op 0 : -> Nat [ctor] . 
  op s : Nat -> Nat [ctor] .  
  sort Config . 
  op <_`,_> : Nat Nat -> Config [ctor] .

  vars R W : Nat .

  rl [enter-w] : < 0, 0 > => < 0, s(0) > . 
  rl [leave-w] : < R, s(W) > => < R, W > . 
  rl [enter-r] : < R, 0 > => < s(R), 0 > . 
  rl [leave-r] : < s(R), W > => < R, W > .
endm)

(mod READERS-WRITERS-PROPS is
  pr READERS-WRITERS .
  pr SYMBOLIC-CHECKER * (sort Nat to Nat') .

  subsort Config < State .

  vars N M : Nat .

  ops one-writer-atmost? exclusion? : -> Prop .

  eq < N, 0       > |= one-writer-atmost? = true [variant] .
  eq < N, s(0)    > |= one-writer-atmost? = true [variant] .
  eq < N, s(s(M)) > |= one-writer-atmost? = false [variant] .
  
  eq < 0,    M    > |= exclusion? = true [variant] .
  eq < N,    0    > |= exclusion? = true [variant] .
  eq < s(N), s(M) > |= exclusion? = false [variant] .
endm)

--- result: no counterexample
(lmc [10] < N:Nat, 0 > |= [] one-writer-atmost? .)

--- result: true
(lfmc < N:Nat, 0 > |= [] one-writer-atmost? .)

--- result: possibly spurious counterexample
(lfmc < N:Nat, 0 > |= <> ~ one-writer-atmost? .)

--- result: false (with counterexample)
(lmc < N:Nat, 0 > |= <> ~ one-writer-atmost? .)

--- result: no counterexample
(lmc [10] < N:Nat, 0 > |= [] exclusion? .)

--- result: true
(lfmc < N:Nat, 0 > |= [] exclusion? .)

