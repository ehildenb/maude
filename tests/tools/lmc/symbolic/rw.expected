Maude> 
	    Full Maude 2.7 June 18th 2014


	    Logical Model Checker January 10th 2012

Introduced module READERS-WRITERS

Introduced module READERS-WRITERS-PROPS

logical model check in READERS-WRITERS-PROPS :
  < N:Nat,0 > |= []one-writer-atmost?
result:
  no counterexample found within bound 10

logical folding model check in READERS-WRITERS-PROPS :
  < N:Nat,0 > |= []one-writer-atmost?
result:
  true (complete with depth 3)

logical folding model check in READERS-WRITERS-PROPS :
  < N:Nat,0 > |= <> ~ one-writer-atmost?
result:
  counterexample found at depth 3
  
prefix
  nil
loop
  {< #1:Nat,0 >,'#1 <- s(#3:Nat); '#2 <- #3:Nat,'leave-r}  

logical model check in READERS-WRITERS-PROPS :
  < N:Nat,0 > |= <> ~ one-writer-atmost?
result:
  counterexample found at depth 3
  
prefix
  nil
loop
  {< #1:Nat,0 >,'#1 <- s(#3:Nat); '#2 <- #3:Nat,'leave-r}  

logical model check in READERS-WRITERS-PROPS :
  < N:Nat,0 > |= []exclusion?
result:
  no counterexample found within bound 10

logical folding model check in READERS-WRITERS-PROPS :
  < N:Nat,0 > |= []exclusion?
result:
  true (complete with depth 3)

Maude> Bye.
